engine/api/src/main/scala/pl/touk/nussknacker/engine/api/typed/supertype/NumberTypesPromotionStrategy.scala

```diff
@@ -1,13 +1,43 @@
 package pl.touk.nussknacker.engine.api.typed.supertype
 
-import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypingResult}
+import java.lang
 
-import scala.collection.mutable
+import org.apache.commons.lang3.ClassUtils
+import pl.touk.nussknacker.engine.api.typed.typing._
 
+trait NumberTypesPromotionStrategy extends Serializable {
 
-trait NumberTypesPromotionStrategy {
+  def promoteSingle(typ: TypingResult): TypingResult = promote(typ, typ)
 
-  def promote(left: Class[_], right: Class[_]): TypingResult
+  def promote(left: TypingResult, right: TypingResult): TypingResult = {
+    (toSingleTypesSet(left), toSingleTypesSet(right)) match {
+      case (Left(Unknown), _) => Unknown
+      case (_, Left(Unknown)) => Unknown
+      case (Right(lSet), Right(rSet)) =>
+        val allCombinations = for {
+          l <- lSet
+          r <- rSet
+        } yield promoteClasses(l.objType.klass, r.objType.klass)
+        Typed(allCombinations)
+    }
+  }
+
+  private def toSingleTypesSet(typ: TypingResult) : Either[Unknown.type, Set[SingleTypingResult]] =
+    typ match {
+      case s: SingleTypingResult => Right(Set(s))
+      case u: TypedUnion => Right(u.possibleTypes)
+      case Unknown => Left(Unknown)
+    }
+
+  final def promoteClasses(left: Class[_], right: Class[_]): TypingResult = {
```

Comment:
Do we want TypingResult or Class/TypedClass here? In mathUtils we in fact rely on the fact that this method returns TypedClass?

Reply:
There is little trick here. For some promotions we can't be sure correct type. During random tests I found out that for power operator, Ints can be promoted to Longs or not. So in this special case this promotion strategy return union of types. So basically `promoteClasses` is rather "internal" method. Main method is `promote(left: TypingResult, right: TypingResult): TypingResult`. I could add additional abstraction level for `NumberTypesPromotionStrategy` for cases where during typing we will got only single case. WDYT?

Reply:
Ok... I see now :) 
We can add one more layer - or just add comment/better handling in https://github.com/TouK/nussknacker/pull/1088/files#diff-7d2e0e302a77fb47e52efb8101378415R72

Reply:
I've added additional abstraction - now we have less downcasting - good point.